src/main/java/org/jboss/logmanager/Logger.java

```diff
@@ -880,45 +883,31 @@ public void logRaw(final ExtLogRecord record) {
     }
 
     /**
-     * Set the resource bundle for this logger. Unlike {@link java.util.logging.Logger#setResourceBundle(ResourceBundle)},
-     * there is no parent search performed for resource bundles by this implementation.
+     * Set the resource bundle for this logger.
      *
      * @param resourceBundle the resource bundle (must not be {@code null})
      */
     @Override
     public void setResourceBundle(ResourceBundle resourceBundle) {
-        Assert.checkNotNullParam("resourceBundle", resourceBundle);
-        LogContext.checkAccess();
-        ResourceBundle old;
-        do {
-            old = this.resourceBundle;
-            if (old != null && !old.getBaseBundleName().equals(resourceBundle.getBaseBundleName())) {
-                throw new IllegalArgumentException("Bundle base name does not match existing bundle");
-            }
-        } while (!resourceBundleUpdater.compareAndSet(this, old, resourceBundle));
+        bundleSet.set(true);
+        setResourceBundle(resourceBundle, true);
     }
 
     /**
-     * Get the resource bundle for this logger. Unlike {@link java.util.logging.Logger#getResourceBundle()},
-     * there is no parent search performed for resource bundles by this implementation.
+     * Get the resource bundle for this logger.
      *
      * @return the resource bundle, or {@code null} if none is configured for this logger
      */
     @Override
     public ResourceBundle getResourceBundle() {
-        return resourceBundle;
-    }
-
-    /**
-     * Get the resource bundle name for this logger. Unlike {@link java.util.logging.Logger#getResourceBundleName()},
-     * there is no parent search performed for resource bundles by this implementation.
-     *
-     * @return the resource bundle, or {@code null} if none is configured for this logger
-     */
-    @Override
-    public String getResourceBundleName() {
-        final ResourceBundle resourceBundle = getResourceBundle();
-        return resourceBundle == null ? null : resourceBundle.getBaseBundleName();
+        ResourceBundle bundle = this.resourceBundle;
+        if (bundleSet.compareAndSet(false, true) && bundle == null) {
+            bundle = super.getResourceBundle();
+            if (bundle != null) {
+                setResourceBundle(bundle, false);
+            }
+        }
+        return bundle;
```

Comment:
This overall is unsound. Our resourceBundle field is already atomic, so there is no need to keep a secondary state; doing so introduces a window where `bundleSet` is true and `bundle` isn't set yet, which is bad.

You probably want a simpler, double-checked pattern:

```java
ResourceBundle bundle = this.bundle;
if (bundle == null) {
    synchronized (this) { // the same lock used by the superclass
        bundle = this.bundle;
        if (bundle == null) {
            bundle = super.getResourceBundle();
            if (bundle == null) {
                return null;
            } 
        }
        this.bundle = bundle;
    }
}
return bundle;
```

Also we do not want to drop `getResourceBundleName()`, at least I don't think so (since it would be inconsistent with what `getResourceBundle()` returns). Since `getResourceBundle()` calls `getResourceBundleName()` though, we might have to examine the caller stack frame to determine what action to take in order to avoid a recursion problem.

But, since the locking characteristics of all of this sucks really bad, maybe it's better to stick closer to the spirit of the existing code, and do our own lazy resource bundle resolution based on the first caller of `getResourceBundle()`.

That would look something like this (off the top of my head, not checked at all):

```java
    public ResourceBundle getResourceBundle() {
        ResourceBundle bundle = resourceBundle;
        ResourceBundle newBundle, witness;
        while (bundle == null) {
            String name = getResourceBundleName();
            if (name == null) {
                return null; // no bundle
            }
            // resolve and populate cache
            ClassLoader cl = /* get caller class loader via StackWalker */;
            try {
                newBundle = ResourceBundle.getBundle(name, Locale.getDefault(), cl);
            } catch (MissingResourceException e) {
                // no good
                return null; // maybe cache a null result with a special tombstone RB impl
            }
            witness = bundleHandle.compareAndExchange(this, bundle, newBundle);
            if (witness == bundle) {
                // success
                return newBundle;
            }
            bundle = witness;
        }
        return bundle;
    }

    public void setResourceBundle(ResourceBundle resourceBundle) {
        // simple way to validate and set the bundle name all at once
        super.setResourceBundle(resourceBundle);
        // success! not race-free but good enough™ since at worst the superclass will have a stale ref
        this.resourceBundle = resourceBundle;
    }
```

And delete our `getResourceBundleName()` in favor of the JDK's, which is definitely good enough.

Reply:
The idea of the of the `bundleSet` is for cases when the `super.getResourceBundle()` returns `null`. If we don't do that, then we'll invoke that for every case, instead of just the first time the the `getResourceBundle()` is invoked. We should be able to do that because we override the `setResourceBundle()` as well.

The removal of the `getResourceBundleName()` was indeed for recursion. However, this method is fast anyway in the `java.util.logging.Logger`. It's the instance variable and just returns the name. It seemed fine to just user the super class version.